test/integration/targets/win_psmodule/tasks/main.yml

```diff
@@ -1,28 +1,97 @@
+# This file is part of Ansible
+
 # test code for the win_psmodule module when using winrm connection
-# (c) 2017, Daniele Lazzari <lazzari@mailup.com>
+# Copyright: (c) 2018, Wojciech Sciesinski <wojciech[at]sciesinski[dot]net>
+# Copyright: (c) 2017, Daniele Lazzari <lazzari@mailup.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 
-# This file is part of Ansible
-#
-# Ansible is free software: you can redistribute it and/or modify
-# it under the terms of the GNU General Public License as published by
-# the Free Software Foundation, either version 3 of the License, or
-# (at your option) any later version.
-#
-# Ansible is distributed in the hope that it will be useful,
-# but WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-# GNU General Public License for more details.
-#
-# You should have received a copy of the GNU General Public License
-# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
-
-
-- name: get facts
-  setup:
-
-- name: Perform integration tests for Powershell 5+
-  when: ansible_powershell_version >= 5
+- name: get PowerShell version
+  win_shell: '$PSVersionTable.PSVersion.Major'
+  register: powershell_major_version
+
+- name: update PackageManagement and PowerShellGet when PowerShell < 5.0
+  when: powershell_major_version.stdout | int < 5
+  block:
+    - name: get the TEMP path
+      win_shell: '$env:TEMP'
+      register: temp_path
+
+    - name: download PackageManagement
+      win_get_url:
+        url: https://download.microsoft.com/download/C/4/1/C41378D4-7F41-4BBE-9D0D-0E4F98585C61/PackageManagement_x64.msi
+        dest: '{{ temp_path.stdout | trim }}\PackageManagement_x64.msi'
+
+    - name: install PackageManagement
+      win_package:
+        path: '{{ temp_path.stdout | trim }}\PackageManagement_x64.msi'
+        state: present
+        creates_path: 'C:\Program Files\WindowsPowerShell\Modules\PowerShellGet\PowerShellGet.psd1'
+
+    - name: create the required folder
+      win_file:
+        path: 'C:\Program Files\PackageManagement\ProviderAssemblies'
+        state: directory
+
+    - name: download nuget
+      win_get_url:
+        url: https://dist.nuget.org/win-x86-commandline/latest/nuget.exe
+        dest: 'C:\Program Files\PackageManagement\ProviderAssemblies\NuGet-anycpu.exe'
+
+    - name: update NuGet provider
+      win_shell: 'Find-PackageProvider -Name Nuget -ForceBootstrap -IncludeDependencies'
+
+    - name: get a random folder path name
+      win_shell: "([System.IO.Path]::GetTempPath()+([System.IO.Path]::GetRandomFileName())).Split('.')[0]"
+      register: random_folder
+
+    - name: create a random folder
+      win_shell: 'New-Item -Path {{ random_folder.stdout | trim }} -ItemType Directory | Out-Null'
+
+    - name: download and save the nevest version of the PackageManagement module from PowerShell Gallery
```

Comment:
I'm not sure I see the need for this, can't we just update the package now that PSGet is installed. Also It's best to use `script` if you have a lot of `win_shell` actions in a test, it takes a lot longer to execute each task when you can do it in 1 script.

Reply:
Initially, I tried to do that as you propose. It doesn't work.

Reply:
Can you go into more detail as to what didn't work. Making these tests efficient is a key part of building the tests. Having 1 task that executes a bunch of commands compared to multiple `win_shell` invocations is a lot more desirable. We are already near the limit of our full run in Shippable.

Reply:
I didn't retest that now but as I remember an issue is related that modules have to update themselves - a chicken/egg case. So a chicken has to be killed (`Remove-Module`) and the next an egg can be laid (`Copy-Item`).